src/main/java/io/wispforest/lavender/structure/StructureTemplate.java

```diff
@@ -163,224 +287,330 @@ public int countValidStates(World world, BlockPos anchor, BlockRotation rotation
-                    var predicate = result.right().get();
+    public Iterator<Pair<BlockPos, BlockStatePredicate>> iterator(BlockRotation rotation) {
+        return new StructureTemplateIterator(this, rotation);
+    }
 
-                    var previewStates = new ArrayList<BlockState>();
-                    predicate.tag().forEach(registryEntry -> {
-                        var block = registryEntry.value();
-                        var state = block.getDefaultState();
+    public static class NestedBlockStatePredicate implements BlockStatePredicate {
+        @NotNull
+        private final BlockStatePredicate[] predicates;
 
-                        for (var propAndValue : predicate.vagueProperties().entrySet()) {
-                            Property prop = block.getStateManager().getProperty(propAndValue.getKey());
-                            if (prop == null) return;
+        @NotNull
+        private final BlockState[] previewStates;
 
-                            Optional<Comparable> value = prop.parse(propAndValue.getValue());
-                            if (value.isEmpty()) return;
+        public NestedBlockStatePredicate(@NotNull BlockStatePredicate[] predicates) {
+            this.predicates = predicates;
+            this.previewStates = Arrays.stream(predicates)
+                    .flatMap((predicate) -> Arrays.stream(predicate.previewBlockstates()))
+                    .toArray(BlockState[]::new);
+        }
 
-                            state = state.with(prop, value.get());
-                        }
+        @NotNull
+        @Override
+        public Result test(@NotNull BlockState state) {
+            boolean hasBlockMatch = false;
+            for (var predicate : this.predicates) {
+                var result = predicate.test(state);
+                if (result == Result.STATE_MATCH)
+                    return Result.STATE_MATCH;
+                else if (result == Result.BLOCK_MATCH)
+                    hasBlockMatch = true;
+            }
 
-                        previewStates.add(state);
-                    });
+            return hasBlockMatch ? Result.BLOCK_MATCH : Result.NO_MATCH;
+        }
 
-                    keys.put(key, new BlockStatePredicate() {
-                        @Override
-                        public BlockState preview() {
-                            if (previewStates.isEmpty()) return Blocks.AIR.getDefaultState();
-                            return previewStates.get((int) (System.currentTimeMillis() / 1000 % previewStates.size()));
-                        }
+        @Override
+        public BlockState[] previewBlockstates() {
+            return this.previewStates;
+        }
+    }
 
-                        @Override
-                        public Result test(BlockState state) {
-                            if (!state.isIn(predicate.tag())) return Result.NO_MATCH;
+    public static class SingleBlockStatePredicate implements BlockStatePredicate {
+        @NotNull
```

Comment:
As above, don't annotate non-nullable stuff as `@NotNull` - just leave it blank. And, also as above, don't put newlines between the fields unless there's at least one group of at least two fields that collectively serve a different purpose than the rest